source/common/stats/thread_local_store.cc

```diff
@@ -75,6 +99,31 @@ void ThreadLocalStoreImpl::shutdownThreading() {
   shutting_down_ = true;
 }
 
+void ThreadLocalStoreImpl::mergeHistograms(PostMergeCb merge_complete_cb) {
+  if (!shutting_down_) {
```

Comment:
mergeHistograms is async.  So I wonder if you should keep track of whether there's one in progress, and skip merging in that case.  I think the behavior if you induce two or three merges in a row before merge_complete_cb, the behavior may be hard to reason about.

In that case it may be sufficient to simply drop the callback; that would seem to work well for the call-site in InstanceImpl::flushStats()

Reply:
I do not think there would be two merge happening in parallel unless the merge takes abnormally long. It is called from flushStats which is invoked for every stats flush interval default value of which is 5 sec. This might happen if the flush interval is configured very low, may be less than sec or few milliseconds.  So I think the tracking complication is not required here.

Reply:
All it takes is for a thread to not get scheduled for a few seconds because the machine is busy, and this will be an issue.  Don't you think it's worth checking?

Reply:
I don't think this can happen with how we're using this method since the timer is only rescheduled in the last line of the `merge_complete_cb`, which is called only after all the merging is done.  (Feel free to correct me if I'm misunderstanding the flow.)

If we imagine that this method may be used in other circumstances, I think a check like this could make sense, but this method has a pretty specific use case.  Maybe instead we could leave out the check, but document how this method should be used (it can only be called again once the passed in callback is called)?  Not super opinionated, but it just sometimes makes code (especially complicated code :) ) less readable when inactive constraints are checked.

Reply:
@mrice32 I think you are probably right, but there are two call-sites to flushStats() from server.cc.  It's not immediately obvious to me that there may never be any other initiators.  That is a private method now.

I think this is a very cheap check to add, to avoid having to think too hard about what very confusing (and possibly crashy) behavior might arise from having two or three concurrent mergeHistogram calls running.  WDYT?

Reply:
@mrice32 After pushed this I also realized that the timer is rescheduled only at the last line of the merge_complete_cb. I left it because it does not harm having this as it is safe and cheap trick as @jmarantz mentioned.

Reply:
That's fair - I think the check makes sense.  But I'm somewhat opposed to the idea of having concurrent merges appear as if they are part of the normal program flow and no-op when they happen.  I would personally prefer to document how the method should be used and `ASSERT(!merge_in_progress_)`.  Anyway, either way is fine, that's just my personal preference.  In either case, I think the behavior should be clearly documented in the declaration.

Reply:
ASSERT would be OK with me.  Is there an Envoy equivalent to
```
if (cond) {
  LOG(DFATAL) << "error message";
}
```
Where in release-builds this will log an error, and in debug builds it will also abort?

And in any case documenting how the methods should be used would be great.

Reply:
ok. I have added the ASSERT and updated the documentation of this method. PTAL.